tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_ops.py

```diff
@@ -540,28 +550,58 @@ def initial_value():
-                                   model_mode=ModelMode.CURRENT_SETTING,
-                                   name=trainable_name)
+      def _create_or_get_trainable(trainable_name):
+        if trainable_name is None:
+          if context.executing_eagerly():
+            raise ValueError(
+                'Must provide a name for embedding_lookup when using eager execution.'
+            )
+          trainable_name = ops.get_default_graph().unique_name(
+              _ANONYMOUS_TRAINABLE_STORE_KEY)
+        if not context.executing_eagerly() and not ops.inside_function():
+          wrapper = de.TrainableWrapper(params,
+                                        ids,
+                                        max_norm=max_norm,
+                                        initial_value=initial_value,
+                                        dtype=params.value_dtype,
+                                        trainable=params.trainable,
+                                        collections=collections,
+                                        model_mode=ModelMode.CURRENT_SETTING,
+                                        name=trainable_name)
+          params._trainable_store[trainable_name] = wrapper
+          return wrapper
+        else:
+          with ops.init_scope():
+            shadow = params._trainable_store.get(trainable_name, None)
+            if shadow is None:
+              shadow = de.shadow_ops.ShadowVariable(
+                  params,
+                  name=trainable_name,
+                  max_norm=max_norm,
+                  trainable=params.trainable,
+                  model_mode=ModelMode.CURRENT_SETTING)
+              params._trainable_store[trainable_name] = shadow
+          return shadow
 
       with ops.colocate_with(ids, ignore_existing=True):
-        if context.executing_eagerly():
-          trainable_ = params._trainable_store.get(name, None)
-          if trainable_ is None:
-            trainable_ = _create_trainable(name)
-            params._trainable_store[name] = trainable_
-          else:
-            trainable_._reset_ids(ids)
-        else:
-          trainable_ = _create_trainable(name)
-          params._trainable_store[name] = trainable_
+        trainable_ = _create_or_get_trainable(name)
+
+      if isinstance(trainable_, de.shadow_ops.ShadowVariable):
+        embeddings = de.shadow_ops.embedding_lookup(
+            trainable_,
+            ids,
+            partition_strategy=partition_strategy,
+            name=name,
+            validate_indices=validate_indices)
+        if return_trainable:
+          if not context.executing_eagerly():
+            raise NotImplementedError(
+                'return_trainable currently is not implemented when using tf.function.'
+                ' Please use `Variable.trainable_store` or `Variable.get_trainable_by_name`'
+                ' to access the shadow trainable variable if call `embedding_lookup` series'
```

Comment:
User maybe don't know `shadow`,  replace with tf.function on Graph Mode?

Reply:
I add some comments on both ShadowVariable and TrainableWrapper to elaborate on their relationship. And in most cases, user will not sense them if use Keras API. And if only if the dynamic_embedding.embedding_lookup series APIs are used inside tf.function scope, the user will need to access the local trainable by `trainable_store` or `get_trainable_by_name`. Added some comments on it.

If some users need to develop new features on that, like customizing a new embedding Keras layer, they could use both `ShadowVariable` or `embedding_lookup`. And `ShadowVariable` could provide the ability to operate on the local trainable buffer directly.

Reply:
> User maybe don't know `shadow`, replace with tf.function on Graph Mode?

And also the `tf.function` is originally disabled in TensorFlow